Add module source address parsing

dbanck opened the pull request:
This PR adds the parsing of module source addresses for local and registry modules. It exposes one new function:
* `ParseRawModuleSourceRegistry`

It's based on the [`internal/addrs/module_source.go`](https://github.com/hashicorp/terraform/blob/main/internal/addrs/module_source.go) code in terraform core with some modifications:
* Removed anything related to `ModuleSourceRemote` (e.g. GitHub or Bitbucket)
* Removed anything related to `ModuleSourceLocal` (e.g. `./example`)
* Removed the go-getter dependency
  * Was required via [`internal/getmodules`](https://github.com/hashicorp/terraform/tree/main/internal/getmodules)
  * Detection was required for `ModuleSourceRemote`
  * Inlined [`SourceDirSubdir`](https://github.com/hashicorp/terraform/blob/0028a26cf81d6b9d7c036407426469597c017a02/internal/addrs/module_source.go#L16)

### Example usage

```go
moduleSourceRegistry, err := tfaddr.ParseRawModuleSourceRegistry(module.SourceAddr)
if err == nil {
	bodySchema.DocsLink = &schema.DocsLink{
		URL: fmt.Sprintf(
			`https://%s/modules/%s/%s`,
			moduleSourceRegistry.PackageAddr.Host.ForDisplay(),
			moduleSourceRegistry.PackageAddr.ForRegistryProtocol(),
			"latest",
		),
	}
}
```
Enables https://github.com/hashicorp/terraform-ls/issues/877

radeksimko reviewed (commented):
This LGTM aside from one question/aspect:

Is there a particular reason for including `ParseRawModuleSource`, `ModuleSourceLocal` and the `ModuleSource` interface? The detection of local modules doesn't seem all that complicated and I _think_ we could get away with it living either in terraform-ls, or potentially under `terraform-schema`: https://github.com/hashicorp/terraform-ls/blob/94aa1d8d66f28441fbf8e3273eefcc5554b9b817/internal/terraform/datadir/module_types.go#L61-L68

I guess there may be a convenience argument for including it, but the risk is that this gives anyone the expectation that the function can parse _any_ valid module source address and I'd really prefer to avoid that scope creep, if possible - esp. because that allows us to avoid go-getter.

_Maybe_ we'll re-scope the library as `hashicorp/terraform-address` at some point, but I'd prefer not to do that unless it's really necessary.

dbanck commented:
Good call! I've removed the code related to local module parsing, and without the `ModuleSource` interface, usage of this library becomes much easier.

I updated the example and removed the type cast.

radeksimko commented on module.go:
```diff
@@ -0,0 +1,268 @@
+// ParseRawModuleSourceRegistry only accepts module registry addresses, and
+// will reject any other address type.
+func ParseRawModuleSourceRegistry(raw string) (ModuleSourceRegistry, error) {
+	// Before we delegate to the "real" function we'll just make sure this
+	// doesn't look like a local source address, so we can return a better
+	// error message for that situation.
+	if isModuleSourceLocal(raw) {
+		return ModuleSourceRegistry{}, fmt.Errorf("can't use local directory %q as a module registry address", raw)
+	}
```
At the risk of drifting a little bit from the (non-public) TF Core implementation, is it worth also removing this check here?

I expect downstream consumers would just check for local addresses 1st anyway.

We can be more defensive if we find edge cases where the parser misinterprets addresses. In this case though it looks like we'd just get slightly different error, which seems okay?

```
--- FAIL: TestParseRawModuleSourceRegistry (0.00s)
    --- FAIL: TestParseRawModuleSourceRegistry/local_path_from_current_dir (0.00s)
        module_test.go:219: wrong error messages
            got:  a module registry source address must have either three or four slash-separated components
            want: can't use local directory "./boop" as a module registry address
    --- FAIL: TestParseRawModuleSourceRegistry/local_path_from_parent_dir (0.00s)
        module_test.go:219: wrong error messages
            got:  a module registry source address must have either three or four slash-separated components
            want: can't use local directory "../boop" as a module registry address
FAIL
FAIL	github.com/hashicorp/terraform-registry-address	0.899s
FAIL
```

dbanck replied:
Good idea, I removed the check and the `isModuleSourceLocal` function.

radeksimko commented on README.md:
````diff
@@ -86,3 +88,31 @@ i.e. assume all of its logic including schema is contained within
+### Example
+
+```go
+local, err := ParseRawModuleSource("./local")
+if err != nil {
+	// deal with error
+}
+
+// local == ModuleSourceLocal("./local")
+
+registry, err := ParseRawModuleSource("hashicorp/subnets/cidr")
````
```suggestion
registry, err := ParseRawModuleSourceRegistry("hashicorp/subnets/cidr")
```

dbanck replied:
Good catch!

radeksimko reviewed (commented):
One more question + docs fix 😅

radeksimko reviewed (approved):
:shipit: 🚀
